### ISSUE TITLE
`Follower(Application)` does not call `self.construct_mapper()`

### ISSUE THREAD

**Issue by wapiflapi** [pyeventsourcing/eventsourcing#261]
I am following the example "Aggregate 8 - Pydantic with declarative syntax".

As soon as I add my custom mapper and transcoder to the applications in a system, I am getting very weird behavior:

In `myapp.policy` I am getting a `domain_event` with satisfies `assert isinstance(domain_event, pydantic.BaseModel)`, BUT it's attributes are still `dict`s, and don't seem to have been parsed by pydantic.

It took me a while but I understood why this it. It's besides the main-point but I'll include it at the end of this issue.

Clearly the mapping isn't happening. I prooved this through logging in my mapper.

### What I think the problem is

Looking at the code in `system.py`:

I think the bug comes from https://github.com/pyeventsourcing/eventsourcing/blob/9.2/eventsourcing/system.py#L85, here is what I understand :

`Follower(Application).follow(...)` "Constructs a notification log reader and a mapper for the named application, and adds them to its collections [...]".

To do so it uses:
```python
factory = self.construct_factory(env)
mapper = factory.mapper(self.construct_transcoder())`
```

This explains why I'm seeing the correct transcoder (using orjson) but not the correct mapping (using pydantic).

### What I think might be a solution

I think this should be either `mapper = self.construct_mapper()` (which is plausible given the use of `self.construct_transcoder()`.

When I change that line, my app works.

A more complicated thing to do  would be to use the mapper of the named application we are following, since that might be a different mapper. I am not sure having different applications use different mappers in the same system is supported, anyway ?

### Why I was seeing such weird behavior

Why am is `assert isinstance(domain_event, pydantic.BaseModel)` ok, BUT it's attributes are still `dict`s, and don't seem to have been parsed by pydantic ?

It is because the default `eventsourcing.persistence.Mapper` does something like this:

```python
    domain_event = object.__new__(cls)
    domain_event.__dict__.update(event_state)
    return domain_event
```

Overwriting pydantic's `__dict__` like that is sure to get it confused :smile:

**Comment by wapiflapi** [comment]
While writing a fix for this on my Application inheriting from ProcessApplication I came up with:

```python
class MyApplication(ProcessApplication):

    # Fixing a bug (?) in eventsourcing.
    def follow(self, name: str, log: NotificationLog) -> None:
        """
        Constructs a notification log reader and a mapper for
        the named application, and adds them to its collections
        of readers and mappers.
        """
        assert isinstance(self.recorder, ProcessRecorder)
        reader = NotificationLogReader(log, section_size=self.pull_section_size)
        env = self.construct_env(name, self.env)
        factory = self.construct_factory(env)
        mapper = factory.mapper(self.construct_transcoder(), type(self.mapper))
        self.readers[name] = reader
        self.mappers[name] = mapper
```

The important line is:

```python
        mapper = factory.mapper(self.construct_transcoder(), type(self.mapper))
```

That might be somewhat cleaner, since it still uses the factory that is built for the `env` related to `name`, instead of our own.

To be honnest I don't understand enough about the implications of this to have an opinion on if this is safe or not but ...

**Comment by johnbywater** [comment]
Hi @wapiflapi thanks for noticing and looking into this issue, and describing it so well.

Sorry for the delay in responding. I was just away on an extended summer break. I'm back now.

The solution you propose, of passing in the mapper class, seems good to me. It's a very simple fix, so long as all the applications are using the same type of mapper. The example doesn't really go as far as you have taken this.

I suppose if we wanted to support applications using different mapper classes, then we could specify the mapper classes as topics in the environment. And then have the factory resolving the topics to mapper classes. With the solution you provided as the default, where they all use the same class.

I've just made the change you suggested in the code, and everything seems to work okay! I will release a new version with this small change, and think about supporting different mapper classes in future.

**Comment by johnbywater** [comment]
Actually, both MAPPER_TOPIC and CIPHER_TOPIC are defined but only CIPHER_TOPIC is used. So I guess I was thinking (in 2021) to implement this but didn't get around to it. I think it would be quite easy to implement this, along the same lines as the support for CIPHER_TOPIC. That is, to read the environment and resolve the topic if defined and otherwise use the default class.

**Comment by johnbywater** [comment]
Changes above released as v9.2.21:
https://pypi.org/project/eventsourcing/9.2.21/
https://eventsourcing.readthedocs.io/en/stable/topics/release_notes.html#version-9-2-21-released-13-oct-2023
